### DISCUSSION TITLE
Implementation Questions

### DISCUSSION THREAD

**Discussion by tsujp** [asmvik/yabai#1296] [General]
Would you prefer such a thread be an Issue or an email-chain instead? I figured a Discussion for others' discovery. I have a few questions popping up as I examine the source more thoroughly and I wanted to get some queries answered.

On [line 23 of the current Makefile](https://github.com/koekeishiya/yabai/blob/32ce2f9590151b37c288cfe5559375bd0e1522e6/makefile#L23) I see `xcrun clang $(OSAX_PATH)/loader.m -shared -O2 -mmacosx-version-min=10.13 -arch x86_64 -o $(OSAX_PATH)/loader -framework Foundation` which is compiling `loader.m` into a dynamic library named `loader`.

On [line 26 of the current Makefile](https://github.com/koekeishiya/yabai/blob/32ce2f9590151b37c288cfe5559375bd0e1522e6/makefile#L26) I see `xxd -i -a $(OSAX_PATH)/loader $(OSAX_PATH)/loader_bin.c` which hexdumps the contents (so effectively "it") of shared library `loader` into `loader_bin.c` which is then compiled later on the Makefile.

Is it done this way because macOS doesn't support static linking against the C runtime (when I tried I get `-lcrt0.o` errors) and so there's no other way to get the `loader` statically linked to `loader_bin`? That's what I gather from this where you're instead straight dumping the object code into `loader_bin.c` before it's compiled resulting in a statically linked library; but I wanted to ask to be sure!

**Comment by asmvik** [comment]
Discussions are fine for this purpose.

So the way this works is that I compile the source for the scripting-addition separately; there are 3 such binary files that are built so far.
After they are built, I use `xxd` to do a dump of the file into `C` arrays, that include the actual bytes of the files. These *dump files* are then built into yabai alongside the regular source. When you run `yabai install-sa` this binary data (for each file/array) is written out to disk at the appropriate location. This is just a way for me to avoid having to ship 4 separate binaries to the user that they then have to place correctly on their system; it's basically just a way to implement a primitive installer -- it has nothing to do with linking really. See #1287 for potential future changes.

The `loader` you reference is the entry point of the scripting-addition infrastructure that macOS supports. This is an old artifact that yabai only uses on High Sierra, Mojave, and Catalina. It is still included for backwards compatibility.

The `mach_loader` is  the new version of `loader` that no longer relies on the scripting-addition infrastructure, but it is simply an executable that performs injection of a shared library, into Dock.app.

The `payload` file is the shared library containing the code that we actually want to inject into the Dock process, and this is where the extended functionality is implemented.
